prusti-encoder/src/encoders/type/ty_pure.rs

```diff
@@ -0,0 +1,249 @@
+
+use prusti_rustc_interface::{
+    abi,
+};
+
+use crate::encoders::{domain::{DomainDataEnum, DomainDataField, DomainDataImmRef, DomainDataMutRef, DomainDataPrim, DomainDataStruct, DomainEnc, DomainEncOutput, DomainEncOutputRef, DomainEncSpecifics}, lifted::{casters::{CastFunctionsOutputRef, CastTypePure}, rust_ty_cast::{GenericCasterPure, RustTyCastersEnc}}};
+
+use super::{
+    most_generic_ty::extract_type_params,
+};
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyPureDataImmRef<'vir> {
+    param: GenericCasterPure<'vir>,
+    data: DomainDataImmRef<'vir>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyPureDataMutRef<'vir> {
+    param: GenericCasterPure<'vir>,
+    data: DomainDataMutRef<'vir>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyPureDataStruct<'a, 'vir> {
+    params: &'a [GenericCasterPure<'vir>],
+    data: DomainDataStruct<'vir>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyPureDataField<'vir> {
+    param: Option<GenericCasterPure<'vir>>,
+    field: DomainDataField<'vir>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TyPureDataEnum<'vir> {
+    data: DomainDataEnum<'vir>,
+}
+
+pub struct TyPureEnc;
+
+#[derive(Clone, Debug)]
+pub struct TyPureEncOutputRef<'vir> {
+    pub snapshot: vir::TypeSnap<'vir>,
+    domain: DomainEncOutputRef<'vir>,
+}
+
+#[derive(Clone, Debug)]
+pub struct TyPureEncOutput<'vir> {
+    inner: TyPureEncOutputRef<'vir>,
+    output: DomainEncOutput<'vir>,
+    params: Vec<GenericCasterPure<'vir>>,
+}
+
+impl<'vir> Deref for TyPureEncOutput<'vir> {
+    type Target = TyPureEncOutputRef<'vir>;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'vir> task_encoder::OutputRefAny for TyPureEncOutputRef<'vir> {}
+
+impl TaskEncoder for TyPureEnc {
+    task_encoder::encoder_cache!(TyPureEnc);
+
+    type TaskDescription<'vir> = ty::Ty<'vir>;
+
+    type OutputRef<'vir> = TyPureEncOutputRef<'vir>;
+    type OutputFullLocal<'vir> = TyPureEncOutput<'vir>;
+
+    type TaskKey<'tcx> = Self::TaskDescription<'tcx>;
+
+    type EncodingError = ();
+
+    fn task_to_key<'vir>(task: &Self::TaskDescription<'vir>) -> Self::TaskKey<'vir> {
+        *task
+    }
+
+    fn do_encode_full<'vir>(
+        task_key: &Self::TaskKey<'vir>,
+        deps: &mut task_encoder::TaskEncoderDependencies<'vir, Self>,
+    ) -> EncodeFullResult<'vir, Self> {
+        with_vcx(|vcx| {
+            let (generic_ty, args) = extract_type_params(vcx.tcx(), *task_key);
+            let domain = deps.require_ref::<DomainEnc>(generic_ty)?;
+            let snapshot = (domain.domain)();
+            let inner = TyPureEncOutputRef { snapshot, domain };
+            deps.emit_output_ref(*task_key, inner.clone())?;
+            let mut params = Vec::new();
+            for arg in args {
+                params.push(deps.require_local::<RustTyCastersEnc<CastTypePure>>(arg)?);
+            }
+            // TODO: mutable references are unsound since they both hold the
```

Comment:
Sounds like we should change this -- why is there permission inside the (direct) predicate at all?